<a href="https://colab.research.google.com/github/Pragna235/Capstone_Python_Projects/blob/main/Bengaluru_House_Prices_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Capstone Project 7: Bengaluru House Prices - Data Cleaning


---

## Instructions

### Goal of the Project:

From class 21 to class 24, you learned the following concepts: 

 - Creating box plots
 - Slicing a DataFrame
 - Treating missing values 

In this project, you will apply what you have learned in class 21 - 24 to achieve the following goals.

|||
|-|-|
|**Main Goal**|Find the missing values in the DataFrame and eliminate the rows with missing values.
|||
|**Additional Goal 1**|Eliminate the columns with missing values that are not required for our analysis|
|||
|**Additional Goal 2**|Replace The missing values and check the impact using boxplot|



---

### Story

A techie, residing in Mumbai, who is a data scientist got an excellent job opportunity from one of his dream company in Bengaluru. He soon wants to shift to Bengaluru and hence, he decides to buy a home in the city.
The lack of trust in property developers in the city resulted in a drop of 7% in the housing units sales across India in 2017. The property prices in Bengaluru fell by around 5% in the second half of 2017, said a study published by a property consultancy firm called Knight Frank. Here's the link to the news article:

[Residential real estate prices drop 3% in big cities
](https://www.businesstoday.in/sectors/infra/residential-real-estate-prices-property-drop-3-per-cent-realty/story/267752.html)

Buying a home, especially in a city like Bengaluru, is a tricky choice. While the major factors are usually the same for all metro cities, there are others to be considered for the Silicon Valley of India. Because of the millennial crowd, vibrant culture, great climate, and a slew of job opportunities in Bengaluru, it is difficult to ascertain the price for a house in the city.
As he is a data scientist he manages to find the dataset for Bengaluru house prices on the internet.

The dataset he found was vast with some useless information and some empty values. He only wants to extract some of the useful information from the dataset.

---

#### Getting Started

Follow the steps described below to solve the project:

1. Click on the link provided below to open the Colab file for this project.
   
   https://colab.research.google.com/drive/1O0W3FE9uV02sVQMkhtGaXPeucqFa6y2r?usp=sharing 

2. Create the duplicate copy of the Colab file. Here are the steps to create the duplicate copy:

    - Click on the **File** menu. A new drop-down list will appear.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/0_file_menu.png' width=500>

    - Click on the **Save a copy in Drive** option. A duplicate copy will get created. It will open up in the new tab on your web browser.

      <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/1_create_colab_duplicate_copy.png' width=500>

     - After creating the duplicate copy of the notebook, please rename it in the **YYYY-MM-DD_StudentName_CapstoneProject7** format. 

3. Now, write your code in the prescribed code cells.

---

### Problem Statement

The dataset acquired by the techie is full of irregularities, incorrect values, and missing values. As a data scientist (or analyst in this context), your task is to clean the dataset given to you. 

This process of preparing data for analysis by removing or modifying data that is incorrect, incomplete, irrelevant, duplicated, or improperly formatted is known as data cleaning.


---

### Dataset Information

This dataset contains several factors that a potential home buyer considers before purchasing a house. Here are the factors:

1. `area_type`: Whether the house has a built-up area or plot area.

2. `availability`: Whether the house is in the ready-to-move state or is there an exact date on which the house will be ready for occupancy.

3. `location`: The name of the locality in Bengaluru where the house is situated.

4. `size`: Whether the house is 1 BHK, 2 BHK, 3 BHK, etc. The term BHK stands for *Bedroom, Hall & Kitchen*.

5. `society`: Name of the society where the house is located.

6. `total_sqft`: The total area of the house in square feet.

7. `bath`: Number of bathrooms in the house.

8. `balcony`: Number of balconies in the house.

9. `price`: Price of the house in lakhs (Indian rupees).


---

### Things To Do

- Create a Pandas DataFrame for the Bengaluru house prices dataset.

- Find the missing values in the DataFrame and replace them with the right missing values. However, if a column has more than 15% missing values then drop (or remove) the column from the DataFrame except for the `location`, `size` and `total_sqft` columns because the house prices are directly dependent on these three crucial parameters. *In fact, the rate of a house is reported as a cost per unit area of the house*.

- Drop (or remove) the rows containing the missing values in the `location`,	`size`, and `total_sqft` columns. Additionally, drop any row which contains nonsensical values in the context of houses.

---

## Specific Tasks to Achieve the Main Goal:

#### 1. Import Modules

First import all the necessary modules to create a DataFrame and box plots.

In [ ]:
# Import the necessary modules.
import pandas as pd


---

#### 2. Load The Dataset

Create a Pandas DataFrame by loading the dataset which is provided in the comma-separated values (CSV) file. Also, display the number of rows and columns in the DataFrame. Here's the link to the dataset:

https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/bengaluru-house-prices/Bengaluru_House_Prices.csv

In [ ]:
# Load the dataset.
banglore_houses_df=pd.read_csv("https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/bengaluru-house-prices/Bengaluru_House_Prices.csv")
banglore_houses_df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[13320 rows x 9 columns]

In [ ]:
# Display the number of rows and columns.
banglore_houses_df.shape

(13320, 9)

---

#### 3. Check For The Missing Values

- Compute the total number of missing values in the DataFrame.

- Compute the percentage of missing values in the DataFrame. If a column has more than 15% missing values then drop (or remove) the column from the DataFrame. 

In [ ]:
# Find the total number of missing values in each column.
missing_values=banglore_houses_df.isnull().sum()
print(missing_values)

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64


**Hint**: You can find the total number of missing values in each column by using the `sum()` function on top of the `isnull()` function.

**Q:** Which column has the most number of missing values?

**A:** Society column has the most number of missing values.

In [ ]:
# Find the percentage of missing values
for i in banglore_houses_df.columns:
  print(i)
  print(((missing_values[i]*100)/(banglore_houses_df.shape[0])))

area_type
0.0
availability
0.0
location
0.0075075075075075074
size
0.12012012012012012
society
41.306306306306304
total_sqft
0.0
bath
0.5480480480480481
balcony
4.572072072072072
price
0.0


**Hint** : You can calculate the percentage of missing values in each column by multiplying the total number of missing values in each column with 100 and dividing the value with the total number of rows in the DataFrame.

**Q:** Which is/are column(s) that need to be removed? Write "Not Applicable" if there are none.

**A:** Society column needs to be removed.


---

### 4. Data Cleaning - Rows

First things first, drop (or remove) the rows containing the missing values in the `location`,	`size`, and `total_sqft` columns because they are the most important parameters in governing the price of a house.

---

#### 4.1 Removing Rows From The `location` Column

You need to remove all the rows containing the missing values in the `location` column.

In [ ]:
# Get all the rows having the missing values in the 'location'.
banglore_houses_df[banglore_houses_df['location'].isnull()==True]

area_type   availability location  ... bath balcony price
568  Super built-up  Area  Ready To Move      NaN  ...  3.0     2.0  86.0

[1 rows x 9 columns]

**Q:** How many rows have the missing values in the `location` column?

**A:**Only one row has missing values  in the location.

In [ ]:
# Total number of rows having the missing values in the 'location' column.
banglore_houses_df['location'].isnull().sum()

1

In [ ]:
# Discard the rows containing the missing values in the 'location' column.
correct_location_df=banglore_houses_df[~(banglore_houses_df['location'].isnull()==True)]
correct_location_df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[13319 rows x 9 columns]

**Hint** : Use `isnull()` function inside the `loc[]` function to locate and remove the rows with missing values from the `location` column.

---

#### 4.2 Removing Rows From The `size` Column

Now, you need to remove all the rows containing the missing values in the `size` column.

In [ ]:
# Get the rows having the missing values in the 'size' column.
correct_location_df[correct_location_df['size'].isnull()==True]

area_type          availability        location  ... bath balcony    price
579    Plot  Area  Immediate Possession  Sarjapur  Road  ...  NaN     NaN   34.185
1775   Plot  Area  Immediate Possession        IVC Road  ...  NaN     NaN  124.000
2264   Plot  Area  Immediate Possession    Banashankari  ...  NaN     NaN  460.000
2809   Plot  Area  Immediate Possession  Sarjapur  Road  ...  NaN     NaN   28.785
2862   Plot  Area  Immediate Possession     Devanahalli  ...  NaN     NaN   46.800
5333   Plot  Area  Immediate Possession     Devanahalli  ...  NaN     NaN  177.115
6423   Plot  Area  Immediate Possession      Whitefield  ...  NaN     NaN   26.730
6636   Plot  Area  Immediate Possession          Jigani  ...  NaN     NaN   25.490
6719   Plot  Area  Immediate Possession         Hoskote  ...  NaN     NaN   28.545
7680   Plot  Area  Immediate Possession    Kasavanhalli  ...  NaN     NaN  400.000
8306   Plot  Area  Immediate Possession          Anekal  ...  NaN     NaN   16.500
8565   Plot  Area  Immediate Possession      Carmelaram  ...  NaN     NaN  120.000
8703   Plot  Area  Immediate Possession     Mysore Road  ...  NaN     NaN   42.300
10634  Plot  Area  Immediate Possession      Rajankunte  ...  NaN     NaN   31.110
11019  Plot  Area  Immediate Possession       Yelahanka  ...  NaN     NaN   12.750
11569  Plot  Area  Immediate Possession      Hosur Road  ...  NaN     NaN    8.440

[16 rows x 9 columns]

**Q:** How many rows have the missing values in the `size` column?

**A:** 

In [ ]:
# Total number of rows having the missing values in the 'size' column.
correct_location_df['size'].isnull().sum()

16

In [ ]:
# Discard the rows containing the missing values in the 'size' column.
correct_size_df=correct_location_df[~(correct_location_df['size'].isnull()==True)]
correct_size_df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[13303 rows x 9 columns]

**Q:** How many rows are there in the DataFrame after removing the rows that need to be removed?

**A:** There are 13303 rows in teh dataframe after removing the rows that needed to be removed.

---

#### 4.3 Removing Rows From The `total_sqft` Column

Now, you need to remove all the rows containing the missing values in the `total_sqft` column.

In [ ]:
# Get the rows having the missing values in the 'total_sqft' column.
correct_size_df[correct_size_df['total_sqft'].isnull()==True]


Empty DataFrame
Columns: [area_type, availability, location, size, society, total_sqft, bath, balcony, price]
Index: []

**Q:** How many rows have the missing values in the `size` column?

**A:** 16 rows have the missing values in the size column.

---

In [ ]:
# Total number of rows having missing values in the 'total_sqft' column.
correct_total_sqft_df['total_sqft'].isnull().sum()

0

In [ ]:
# Discard all the rows having missing values in the 'total_sqft' column.
correct_total_sqft_df=correct_size_df[~(correct_size_df['total_sqft'].isnull()==True)]
correct_total_sqft_df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[13303 rows x 9 columns]

#### 4.3 Removing Rows From The `bath` Column

Assume that in a practical situation, no house, even in a city like Bengaluru, will have more than 5 bathrooms. Hence, drop all the rows in the `bath` column containing more than 5 bathrooms.

In [ ]:
# Get the rows having more than 5 bathrooms in the 'bath' column.
correct_total_sqft_df[correct_total_sqft_df['bath']>5]


area_type   availability  ... balcony  price
9          Plot  Area  Ready To Move  ...     NaN  370.0
45         Plot  Area  Ready To Move  ...     NaN  200.0
64     Built-up  Area  Ready To Move  ...     0.0  140.0
68         Plot  Area  Ready To Move  ...     0.0   85.0
79     Built-up  Area  Ready To Move  ...     3.0  250.0
...               ...            ...  ...     ...    ...
13221      Plot  Area  Ready To Move  ...     1.0   75.0
13226  Built-up  Area  Ready To Move  ...     3.0  215.0
13277      Plot  Area  Ready To Move  ...     NaN  218.0
13300      Plot  Area  Ready To Move  ...     2.0  145.0
13311      Plot  Area  Ready To Move  ...     2.0  250.0

[515 rows x 9 columns]

**Q:** How many rows have more than 5 bathrooms in the `bath` column?

**A:** 515 rows have more than 5 bathrooms in the bath column.

In [ ]:
# Discard the rows having more than 5 bathrooms in the 'bath' column.
correct_bath_df=correct_total_sqft_df[~(correct_total_sqft_df['bath']>5)]
correct_bath_df

area_type   availability  ... balcony   price
0      Super built-up  Area         19-Dec  ...     1.0   39.07
1                Plot  Area  Ready To Move  ...     3.0  120.00
2            Built-up  Area  Ready To Move  ...     3.0   62.00
3      Super built-up  Area  Ready To Move  ...     1.0   95.00
4      Super built-up  Area  Ready To Move  ...     1.0   51.00
...                     ...            ...  ...     ...     ...
13315        Built-up  Area  Ready To Move  ...     0.0  231.00
13316  Super built-up  Area  Ready To Move  ...     NaN  400.00
13317        Built-up  Area  Ready To Move  ...     1.0   60.00
13318  Super built-up  Area         18-Jun  ...     1.0  488.00
13319  Super built-up  Area  Ready To Move  ...     1.0   17.00

[12788 rows x 9 columns]

**Q:** How many rows have atmost 5 bathrooms in the `bath` column?

**A:** 12788 rows have atmost 5 bathrooms in teh bath.

---

### Submitting the Project

Follow the steps described below to submit the project.

1. After finishing the project, click on the **Share** button on the top right corner of the notebook. A new dialog box will appear.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/2_share_button.png' width=500>

2. In the dialog box, click on the **Copy link** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/3_copy_link.png' width=500>


3. The link of the duplicate copy (named as **YYYY-MM-DD_StudentName_CapstoneProject7**) of the notebook will get copied 

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/4_copy_link_confirmation.png' width=500>

4. Go to your dashboard and click on the **My Projects** option.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/5_student_dashboard.png' width=800>

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/6_my_projects.png' width=800>

5. Click on the **View Project** button for the project you want to submit.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/7_view_project.png' width=800>

6. Click on the **Submit Project Here** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/8_submit_project.png' width=800>

7. Paste the link to the project file named as **YYYY-MM-DD_StudentName_CapstoneProject7** in the URL box and then click on the **Submit** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/9_enter_project_url.png' width=800>


---

## Additional Goal 1

### 1. Data Cleaning - Columns 

Compute the percentage of missing values in the DataFrame again. If a column has more than 15% missing values then drop (or remove) the column from the DataFrame. 


In [ ]:
# Percentage of missing values.


**Hint**: You can calculate the percentage of missing values in each column by multiplying the number of missing values with 100 and dividing the resulting value with total number of rows in the DataFrame.

**Q:** Which is/are column(s) that need to be removed? Write "Not Applicable" if there are none.

**A:** 


**Q:** Did the overall percentage of the missing values drop if compared to the initial percentage of the missing values? Write yes or no.

**A:** 

---

#### 1.1 Columns To Be Retained

Create a list of columns that need to be retained. 

In [ ]:
# List of the columns to be retained.   


**Hint**: You need to create a list of columns that have missing values not more than 15%. This can be done by appending such column names in a list using a `for` loop.

---

#### 1.2 Retaining Columns

Retain appropriate columns in the DataFrame.

In [ ]:
# Retain the appropriate columns in the DataFrame.


**Hint**: You need to create a DataFrame with columns having missing values not more than 15%. This can be done by passing the variable holding the list of columns inside a `loc[]` function. 

Again calculate the percetange of missing values in the DataFrame.

In [ ]:
# Percentage of the missing values in the DataFrame.


**Q:** Write down (or identify) the column(s) in which the values need to be replaced with an appropriate value.

**A:** 

---

**SAVE** all the changes made to the project and **SUBMIT** the shareable link in the Student Dashboard Projects panel against the correct class number.


---

## Additional Goal 2

#### 1 Replace The Missing Values 

While performing this task, you need to make sure that the replaced values do not drastically affect the trend of the values in the identified column(s) from the entire dataset.

You can check the trend of the values by creating a box plot before and after replacing the missing (or null) values with the appropriate values.

To do this exercise, you should follow the steps described below:

1. Find out the nature of the values in the identified column(s). 

    - If the value is a continuous value like a price then replace the missing values with the median value.

    - If the value is a discrete value like the number of bedrooms or the number of bathrooms then replace the missing values with either the modal value or the median value. Make sure the replacement value makes sense in the real-world context. For e.g., there can't be 2.5 bathrooms or 2.5 balconies in a house. Also, the replacement value should make sense in the context of the house configuration. For e.g., a 2 BHK house having 5 bathrooms doesn't make sense. Similarly, a 2 BHK house having 5 balconies doesn't make sense.

    - Get the descriptive statistics of the identified column(s) to understand the trend of values.

    - Create a box plot to understand the trend of the values.

2. Replace the values in the identified column(s).

3. Get the descriptive statistics of the identified column(s) and also create a box plot to ensure that the replaced values do not drastically affect the trend of the values.

In [ ]:
# Get the descriptive statistics of the identified column(s).


In [ ]:
# Create boxplot before replacing the missing values in the identified column(s).


**Hint**: You can create a boxplot column by passing the column name `balcony` and the DataFrame inside the `boxplot()` function.

In [ ]:
# Computing the modal value in the 'balcony' column.


**Q:** What should be the replacement value and why?

**A:** 

Now, let's replace the missing values with the appropriate replacement value. But first, create a list of row indices of the rows containing the missing values in the identified column(s).

In [ ]:
# Create a list of indices of the rows containing the missing values in the identified column(s).


**Hint**: You can get the indices of the rows with missing values by using the `index` keyword on top of the `isnull()` function.

In [ ]:
# Replace the missing values in the identified column(s) with the appropriate value.


In [ ]:
# Display the first 5 rows of the DataFrame, after replacing the missing values in the identified column(s).


In [ ]:
# Check for the missing values again.


**Q:** Are there any more missing values in the DataFrame? Write yes or no. Also, write down the next steps that you will take to ensure that the replacement does not drastically affect the trend of values in the identified column(s).

**A:** 

In [ ]:
# Create a box plot after replacing the missing values.


In [ ]:
# Get the descriptive statistics.


**Q:** Do you see any drastic change in the trend of the values in the identified column(s)? Write either yes or no. Justify your answer as well.

**A:** 

---

#### 1.2 Convert Data-Types

Convert the values in the `bath` and `balcony` columns to integer values.

In [ ]:
# Convert the values in the 'bath' and 'balcony' columns to integer values.


In [ ]:
# Print the data-types of the values in the 'bath' and 'balcony' columns.


In [ ]:
# Display the first 5 rows of the DataFrame.


---

**SAVE** all the changes made to the project and **SUBMIT** the shareable link in the Student Dashboard Projects panel against the correct class number.


***REMEMBER… Every step is an opportunity toward success!!!** 
After submitting your project your teacher will send you feedback on your work.



---